### PR TITLE
[READY] Search user defined python binary also in PATH

### PR DESCRIPTION
**Pull request by vheon**
I'm marking as an RFC because there is also #424 which does almost the same thing.

---
This change is [Review on Reviewable](https://reviewable.io/reviews/valloric/ycmd/429)

### CODE REVIEW

**Comment by Valloric** [comment]
Could you summarize the difference between this and #424 for those of us without the full context?

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[ycmd/utils.py, line 190 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KCgkfs2HnF6gKVpbXSD:-KCgkfs3yHe8tPXANfb5:-245419700) ([raw file](https://github.com/valloric/ycmd/blob/95af122ffc7a82fe65bd4c9ce70643efd74d1b14/ycmd/utils.py#L190)):*
I strongly prefer `FindExecutable`. `Which` is a magic name that references the person's prior knowledge of bash, which the code reader not have.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/429#-:-KCglHtPJXhsLgLsy4WA:595133933)*

**Comment by vheon** [comment]
Actually this is basically what we've discussed with @allangarret on #424, but I was already experimenting with this approach and I thought that instead than trying out, explaining to him what to do and check I just opened a PR myself.

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/429#-:-KCgpGyvET3RecuKq2Nz:122980628)*

**Comment by vheon** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[ycmd/utils.py, line 190 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KCgkfs2HnF6gKVpbXSD:-KCgqqgD7b-xWuvZ--Ze:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/95af122ffc7a82fe65bd4c9ce70643efd74d1b14/ycmd/utils.py#L190)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/429)*

**Comment by Valloric** [comment]
I understand that this PR is a result of a prior discussion, but that's spread out, involves reading the code in the other PR etc. Writing a couple of sentences in this PR describing the intent makes things clearer both for current reviewers and people reaching this PR by going through git history.

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[ycmd/utils.py, line 39 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KCh8HN41TVGd7rLSo3K:-KCh8HN41TVGd7rLSo3L:-1005276367) ([raw file](https://github.com/valloric/ycmd/blob/188883955c76407c1ffd4ac02295e5d4f9a61ced/ycmd/utils.py#L39)):*
@micbou Thoughts on the above comment?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/429#-:-KCh8SQ__JO5M39BLDBB:1046106626)*

**Comment by vheon** [comment]
As of right now is possible to specify which Python interpreter to use when spawning a JediHTTP instance by setting the `python_binary_path` option, but that option only accept absolute path to a Python binary. With this PR the behaviour is the following:

- If no option is specified the default is `sys.executable` (as before)
- If a full path is specified, then is used as the Python interpreter to spawn JediHTTP (as before)
- If an executable name is specified then is searched through the user's PATH and the first found is used to spawn a JediHTTP instance

To achieve this I've adapted out `FindExecutable` to  [shutil.which](https://hg.python.org/cpython/file/3.5/Lib/shutil.py#l1081)

The only doubt that I still have is if we want to check that what is passed is really a Python interpreter, similar to what YCM does before spawning `ycmd`. Thoughts?

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/429#-:-KChFl_81fMuL1T0Zdeb:-1328609199)*

**Comment by Valloric** [comment]
Thanks for the clarification. I think this approach is good.

I don't think we need to check is it a Python interpreter; we do that in YCM because `sys.executable` inside vim points to the vim executable, and calling that as Python is bad. For the option, if the user gives us a path to something that's not a Python, it's their fault.

I'm :lgtm: on this, though I would like to see some tests for this. It will require quite a bit of mocking though.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[ycmd/utils.py, line 191 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KChO2_5AnH-AjRfnz8U:-KChO2_5AnH-AjRfnz8V:-707792930) ([raw file](https://github.com/valloric/ycmd/blob/188883955c76407c1ffd4ac02295e5d4f9a61ced/ycmd/utils.py#L191)):*
Do we need this empty line?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/429#-:-KChNYLQy6Oigd-RitiP:-2012786889)*

**Comment by micbou** [comment]
I tested it in a virtualenv on Windows and it was working so :LGTM: with minor comments.

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[ycmd/utils.py, line 39 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KCh8HN41TVGd7rLSo3K:-KChSQfZBjUs9sBslZ72:-770090665) ([raw file](https://github.com/valloric/ycmd/blob/188883955c76407c1ffd4ac02295e5d4f9a61ced/ycmd/utils.py#L39)):*
Since it's done in `shutil.which`, I think we should.

---

*[ycmd/utils.py, line 188 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KChSughY3P2Rd4h2ii1:-KChSugiwVHvpcuMs4Bl:-1608216173) ([raw file](https://github.com/valloric/ycmd/blob/188883955c76407c1ffd4ac02295e5d4f9a61ced/ycmd/utils.py#L188)):*
This comment should be changed saying that we are backporting `which` from `shutil` to Python 2.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/429#-:-KCkCgdGVussV9_9LN2b:-1249044609)*

**Comment by allangarret** [comment]
I tested on archlinux using virtualenvs for python2 and python3 and the autocompletion works out of the box with third party packages with `let g:ycm_python_binary_path = 'python'` inside `vimrc`.

I think we should add a note inside YouCompleteMe documentation about this setting and virtualenv usage, because it will solve, at least, issues [587](https://github.com/Valloric/YouCompleteMe/issues/587) and [609](https://github.com/Valloric/YouCompleteMe/issues/609).

Please, feel free to close #424 and merge this instead, so :lgtm: on this

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/429#-:-KCl8akbzL5QNh3dYRnf:-156779630)*

**Comment by allangarret** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[ycmd/utils.py, line 191 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KChO2_5AnH-AjRfnz8U:-KClB_e8sLraXJRnahDb:1248519265) ([raw file](https://github.com/valloric/ycmd/blob/188883955c76407c1ffd4ac02295e5d4f9a61ced/ycmd/utils.py#L191)):*
It's not necessary. In fact, in another function inside this source file there is no empty line after the declaration although the function starts with a comment

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/429)*

**Comment by vheon** [comment]
> I think we should add a note inside YouCompleteMe documentation about this setting and virtualenv usage, because it will solve, at least, issues 587 and 609.

Once I finish to write the tests and this gets merged in I will send a PR on YCM for updating the docs :+1: 

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/429#-:-KCllJexJJDTay5QAMAw:-1991254707)*

**Comment by Valloric** [comment]
What's the state of this PR now?

**Comment by vheon** [comment]
> What's the state of this PR now?

I actually forgot about this (shame on me 😭 ). I was on a short vacation, so I'll resume this work ASAP.

**Comment by Valloric** [comment]
> so I'll resume this work ASAP

No need for the ASAP part; take your time. :)

**Comment by vheon** [comment]
I've updated the jedi_completer tests and added a couple one for `FindExecutable`.

**Comment by micbou** [comment]
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[ycmd/utils.py, line 223 \[r4\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KI3LW50PgxYm_p4B4ZG:-KI3LW50PgxYm_p4B4ZH:-1882655054) ([raw file](https://github.com/valloric/ycmd/blob/90d0ad31964e3265fcce4756571b9f39af1ff5ab/ycmd/utils.py#L223)):*
> ```Python
> # to be backward compatible with Python2 and more consistent to our codebase.
> def FindExecutable( executable ):
> ```

There is still a blank line here. Not important though.

---

*[ycmd/utils.py, line 246 \[r4\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KI3L4E6aFjKhQrQaDe4:-KI3L4E6aFjKhQrQaDe5:-1282195256) ([raw file](https://github.com/valloric/ycmd/blob/90d0ad31964e3265fcce4756571b9f39af1ff5ab/ycmd/utils.py#L246)):*
> ```Python
>       files = [ executable ]
>     else:
>       files = [ base + ext for ext in WIN_EXECUTABLE_EXTS ]
> ```

`WIN_EXECUTABLE_EXTS` constant does not exist anymore. We need to use the `PATHEXT` environment variable here. This should fix the AppVeyor runs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/429)*

**Comment by vheon** [comment]
Review status: 4 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[ycmd/utils.py, line 223 \[r4\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KI3LW50PgxYm_p4B4ZG:-KI4vPA844iBBGj0ReES:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/90d0ad31964e3265fcce4756571b9f39af1ff5ab/ycmd/utils.py#L223)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> There is still a blank line here. Not important though.
</details>
Done.

---

*[ycmd/utils.py, line 246 \[r4\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KI3L4E6aFjKhQrQaDe4:-KI4vPn6wJeBBsfp0D7a:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/90d0ad31964e3265fcce4756571b9f39af1ff5ab/ycmd/utils.py#L246)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> `WIN_EXECUTABLE_EXTS` constant does not exist anymore. We need to use the `PATHEXT` environment variable here. This should fix the AppVeyor runs.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/429)*

**Comment by vheon** [comment]
Anything else to do here? the coverage decreased because more Windows code was added and coveralls counts only the linux build (maybe would be actually helpful to look at codecov.io...)

**Comment by micbou** [comment]
Tests are not passing on Windows.

**Comment by micbou** [comment]
@vheon Did you read my messages on gitter? I see that you are struggling to make it work on Windows. Don't bother and use [my changes](https://github.com/micbou/ycmd/commit/4bfdb31d0b7f171a6956d9771b0f55c9ca4f7d74). I can also send my changes against your PR if you want.

**Comment by vheon** [comment]
> @vheon Did you read my messages on gitter? I see that you are struggling to make it work on Windows. Don't bother and use my changes. I can also send my changes against your PR if you want.

No I didn't saw it, but I guess that I'm close enough to at least see if I can make it work 😝

**Comment by vheon** [comment]
Now I think I got it. In the end I made what @micbou already did, but oh well 😝  For the coverage drop I'm looking at codecov, since the coverage went down because more windows-only code was needed for windows 😒

**Comment by micbou** [comment]
Do you want to wait for codecov to merge this PR?

<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> Now I think I got it. In the end I made what @micbou already did, but oh well 😝  For the coverage drop I'm looking at codecov, since the coverage went down because more windows-only code was needed for windows 😒 
</details>


---

Reviewed 1 of 1 files at r7, 1 of 5 files at r14, 1 of 3 files at r15, 2 of 2 files at r16.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[ycmd/utils.py, line 222 \[r16\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KJZzsFHD9zg4xjw08g6:-KJZzsFIrhi8TeVyKVEh:-1355005344) ([raw file](https://github.com/valloric/ycmd/blob/cef879945fe802b42b92b2aa4ae71ceca8082d2a/ycmd/utils.py#L222)):*
> ```Python
> 
>   for exe in _GetPossibleWindowsExecutable( filename ):
>     if os.path.exists( exe ) and not os.path.isdir( exe ):
> ```

These two conditions can be replaced by `os.path.isfile( exe )`.

---

*[ycmd/utils.py, line 235 \[r16\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KJ_-0aYleOPlXySRNoP:-KJ_-0aYleOPlXySRNoQ:1540579436) ([raw file](https://github.com/valloric/ycmd/blob/cef879945fe802b42b92b2aa4ae71ceca8082d2a/ycmd/utils.py#L235)):*
> ```Python
> 
>   # Directories usually have execute permission so we need to exclude them.
>   if ( not os.path.isdir( filename )
> ```

This can also be replaced by `os.path.isfile( filename )`. In fact, it is better because it checks for existence too.

---

*[ycmd/completers/typescript/typescript_completer.py, line 102 \[r16\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KJ_00MFlPFpVUiW-tD5:-KJ_00MFlPFpVUiW-tD6:-1908461988) ([raw file](https://github.com/valloric/ycmd/blob/cef879945fe802b42b92b2aa4ae71ceca8082d2a/ycmd/completers/typescript/typescript_completer.py#L102)):*
> ```Python
>       _logger.error( BINARY_NOT_FOUND_MESSAGE )
>       raise RuntimeError( BINARY_NOT_FOUND_MESSAGE )
>     _logger.info( 'found tsserver at {0}'.format( binarypath ) )
> ```

Capital letter?

---

*[ycmd/tests/utils_test.py, line 479 \[r16\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KJ_0ZUZtT7g8sGp5VgH:-KJ_0ZUZtT7g8sGp5VgI:-770960683) ([raw file](https://github.com/valloric/ycmd/blob/cef879945fe802b42b92b2aa4ae71ceca8082d2a/ycmd/tests/utils_test.py#L479)):*
> ```Python
>   with TemporaryExecutable() as executable:
>     dirname, exename = os.path.split( executable )
>     relative_executable = os.path.join( '.', exename )
> ```

Doesn't matter much but `'.'` can be replaced by `os.curdir`.

---

*[ycmd/tests/utils_test.py, line 484 \[r16\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KJ_1SBCEVLL33NCE9sr:-KJ_1SBCEVLL33NCE9ss:695762305) ([raw file](https://github.com/valloric/ycmd/blob/cef879945fe802b42b92b2aa4ae71ceca8082d2a/ycmd/tests/utils_test.py#L484)):*
> ```Python
> 
> 
> @patch.dict( 'os.environ', { 'PATH': tempfile.gettempdir() } )
> ```

Those `patch.dict` are nice.

---

*[ycmd/tests/utils_test.py, line 491 \[r16\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KJ_1HPeQsrNwEV2Lr86:-KJ_1HPeQsrNwEV2Lr87:1275911165) ([raw file](https://github.com/valloric/ycmd/blob/cef879945fe802b42b92b2aa4ae71ceca8082d2a/ycmd/tests/utils_test.py#L491)):*
> ```Python
> 
> 
> def FindExecutable_ReturnNone_IfFileIsNotExecutable_test():
> ```

The underscore between `ReturnNone` and `IfFileIsNotExecutable` seems superfluous.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/429#-:-KJZzhjKTYSP0UZ1GNAI:-315311031)*

**Comment by vheon** [comment]
It would actually be nice to have codecov right away but since it simplify the use of Jedi for virtualenv I'm not sure :confused: What do you think? 

<details><summary><i>Previously, micbou wrote…</i></summary>
> Do you want to wait for codecov to merge this PR?
</details>


---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[ycmd/utils.py, line 222 \[r16\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KJZzsFHD9zg4xjw08g6:-KJbfySOQcovh3EPWI5x:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/cef879945fe802b42b92b2aa4ae71ceca8082d2a/ycmd/utils.py#L222)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> These two conditions can be replaced by `os.path.isfile( exe )`.
</details>
Done.

---

*[ycmd/utils.py, line 235 \[r16\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KJ_-0aYleOPlXySRNoP:-KJbfz6yxx90RjxN36J7:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/cef879945fe802b42b92b2aa4ae71ceca8082d2a/ycmd/utils.py#L235)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> This can also be replaced by `os.path.isfile( filename )`. In fact, it is better because it checks for existence too.
</details>
Done.

---

*[ycmd/completers/typescript/typescript_completer.py, line 102 \[r16\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KJ_00MFlPFpVUiW-tD5:-KJbfzyjNrco20WrssEk:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/cef879945fe802b42b92b2aa4ae71ceca8082d2a/ycmd/completers/typescript/typescript_completer.py#L102)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> Capital letter?
</details>
Done.

---

*[ycmd/tests/utils_test.py, line 479 \[r16\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KJ_0ZUZtT7g8sGp5VgH:-KJbg-pJV1FneqzCTDu7:-1262048144) ([raw file](https://github.com/valloric/ycmd/blob/cef879945fe802b42b92b2aa4ae71ceca8082d2a/ycmd/tests/utils_test.py#L479)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> Doesn't matter much but `'.'` can be replaced by `os.curdir`.
</details>
I actually prefer this form since it seems clear (at least to me) that I want to pass something like `./executable`.

---

*[ycmd/tests/utils_test.py, line 491 \[r16\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KJ_1HPeQsrNwEV2Lr86:-KJbgAHxwvOPuHw_LvgM:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/cef879945fe802b42b92b2aa4ae71ceca8082d2a/ycmd/tests/utils_test.py#L491)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> The underscore between `ReturnNone` and `IfFileIsNotExecutable` seems superfluous.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/429#-:-KJbfLnCTGBLZXofeai7:-271295380)*

**Comment by micbou** [comment]
Let's merge this. Adding coverage for Windows is less important.

@homu r+

<details><summary><i>Previously, coveralls wrote…</i></summary>
> 
> [Coverage Status](https://coveralls.io/builds/6486025)
> 
> Coverage decreased (-0.2%) to 84.736% when pulling **4268395ea11e92c8a7187e3960a1cbd6b74dace3 on vheon:python-binary-from-path** into **e00b7ce7525f99b39d7a75c0ad68f22e83950237 on Valloric:master**.
> 
</details>


---

Reviewed 3 of 3 files at r17.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[ycmd/tests/utils_test.py, line 479 \[r16\]](https://reviewable.io:443/reviews/valloric/ycmd/429#-KJ_0ZUZtT7g8sGp5VgH:-KJbx_NY_i7QkdSDHX2t:-1530771663) ([raw file](https://github.com/valloric/ycmd/blob/cef879945fe802b42b92b2aa4ae71ceca8082d2a/ycmd/tests/utils_test.py#L479)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> I actually prefer this form since it seems clear (at least to me) that I want to pass something like `./executable`.
</details>
Agreed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/429#-:-KJbxlYZD_1W3I0DyI5U:1325433994)*

**Comment by micbou** [comment]
We need to document this feature in YCM README.
